processing/src/main/java/io/druid/segment/ReferenceCountingSegment.java

```diff
@@ -44,141 +63,74 @@ public ReferenceCountingSegment(Segment baseSegment)
 
   public Segment getBaseSegment()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment;
-    }
+    return !isClosed() ? baseSegment : null;
   }
 
   public int getNumReferences()
   {
-    return numReferences;
+    return Math.max(referents.getRegisteredParties() - 1, 0);
   }
 
   public boolean isClosed()
   {
-    return isClosed;
+    return referents.isTerminated();
   }
 
   @Override
   public String getIdentifier()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment.getIdentifier();
-    }
+    return !isClosed() ? baseSegment.getIdentifier() : null;
   }
 
   @Override
   public Interval getDataInterval()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment.getDataInterval();
-    }
+    return !isClosed() ? baseSegment.getDataInterval() : null;
   }
 
   @Override
   public QueryableIndex asQueryableIndex()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment.asQueryableIndex();
-    }
+    return !isClosed() ? baseSegment.asQueryableIndex() : null;
   }
 
   @Override
   public StorageAdapter asStorageAdapter()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        return null;
-      }
-
-      return baseSegment.asStorageAdapter();
-    }
+    return !isClosed() ? baseSegment.asStorageAdapter() : null;
   }
 
   @Override
-  public void close() throws IOException
+  public void close()
   {
-    synchronized (lock) {
-      if (isClosed) {
-        log.info("Failed to close, %s is closed already", baseSegment.getIdentifier());
-        return;
-      }
-
-      if (numReferences > 0) {
-        log.info("%d references to %s still exist. Decrementing.", numReferences, baseSegment.getIdentifier());
-
-        decrement();
-      } else {
-        log.info("Closing %s", baseSegment.getIdentifier());
-        innerClose();
-      }
+    if (closed.compareAndSet(false, true)) {
+      referents.arriveAndDeregister();
```

Comment:
Maybe do we need to check the number of unarrived parties here to avoid the case that decrement() is not called properly?

Reply:
Not sure, what do you suggest here?

Reply:
I mean, is it currently possible to check that close() is called before all references of ReferenceCountingSegment calls decrement()?
I guess that close() is called before decrement() means that someone is still using that segment and should be avoided.

Reply:
It was allowed in the previous version of the code as well. I think the idea is to have something a-la `shared_ptr`

Reply:
Ok, then would you please leave a comment?

Reply:
Sure, added comments